parcels/interaction/base_neighbor.py

```diff
@@ -0,0 +1,126 @@
+from abc import ABC, abstractmethod
+
+import numpy as np
+from parcels.interaction.distance_utils import spherical_distance
+
+
+class BaseNeighborSearch(ABC):
+    """Base class for searching particles in the neighborhood."""
+    name = "unknown"
+
+    def __init__(self, interaction_distance, interaction_depth,
+                 max_depth=100000):
+        self.interaction_depth = interaction_depth
+        self.interaction_distance = interaction_distance
+        self.inter_dist = np.array(
+            [self.interaction_distance, self.interaction_distance,
+             self.interaction_depth]).reshape(3, 1)
+        self.max_depth = max_depth
+        self._values = None
+        self._active_mask = None
+
+    @abstractmethod
+    def find_neighbors_by_coor(self, coor):
+        '''Get the neighbors around a certain location.
+
+        :param coor: Numpy array with [lat, long, depth].
+        :returns List of particle indices.
+        '''
+        raise NotImplementedError
+
+    def find_neighbors_by_idx(self, particle_idx):
+        '''Get the neighbors around a certain particle.
+
+        Mainly useful for Structure of Array (SoA) datastructure
+
+        :param particle_idx: index of the particle (SoA).
+        :returns List of particle indices
+        '''
+        coor = self._values[:, particle_idx].reshape(3, 1)
+        return self.find_neighbors_by_coor(coor)
+
+    def update_values(self, new_values, new_active_mask=None):
+        '''Update the coordinates of the particles.
+
+        This is a default implementation simply rebuilds the structure.
+        If the rebuilding is slow, a faster implementation can be provided.
+
+        :param new_values: new coordinates of the particles.
+        :param new_active_mask: boolean array indicating active particles.
+        '''
+        self.rebuild(new_values, new_active_mask)
+
+    def rebuild(self, values, active_mask=-1):
+        """Rebuild the neighbor structure from scratch.
+
+        :param values: new coordinates of the particles.
+        :param active_mask: boolean array indicating active particles.
+        """
+        if values is not None:
+            self._values = values
+        if active_mask is None:
+            self._active_mask = np.arange(self._values.shape[1])
+
+        # If active_mask == -1, then don't update the active mask.
+        if not (isinstance(active_mask, int) and active_mask == -1):
+            self._active_mask = active_mask
+        self._active_idx = self.active_idx
+
+    @property
+    def active_idx(self):
+        "Indices of the currently active mask."
+        if self._active_mask is None:
+            return np.arange(self._values.shape[1])
+        return np.where(self._active_mask)[0]
+
+    @abstractmethod
+    def _distance(self, coor, subset_idx):
+        """Distance between a coordinate and particles
+
+        Distance depends on the mesh (spherical/flat).
```

Comment:
ah, looking back at my comment from the specific interaction kernels in `application_kernels`, I now imagine then that the point of which distance matrix works on which mesh is guided by the specialisation class of this `BaseNeighborSearch`. Sounds tenable.